plugins/JsTrackerInstallCheck/JsTrackerInstallCheck.php

```diff
@@ -0,0 +1,149 @@
+<?php
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+
+namespace Piwik\Plugins\JsTrackerInstallCheck;
+
+use Piwik\Common;
+use Piwik\Container\StaticContainer;
+use Piwik\Date;
+use Piwik\Log\LoggerInterface;
+use Piwik\Option;
+use Piwik\Plugins\SitesManager\API as SitesManagerApi;
+use Piwik\SettingsPiwik;
+use Piwik\Tracker\Request;
+
+class JsTrackerInstallCheck extends \Piwik\Plugin
+{
+    const QUERY_PARAM_NAME = 'tracker_install_check';
+    const OPTION_NAME_PREFIX = 'JsTrackerInstallCheck_';
+
+    public function registerEvents()
+    {
+        return [
+            'Tracker.isExcludedVisit' => 'isExcludedVisit',
+            'Translate.getClientSideTranslationKeys' => 'getClientSideTranslationKeys',
+            'AssetManager.getStylesheetFiles' => 'getStylesheetFiles',
+        ];
+    }
+
+    public function getClientSideTranslationKeys(&$translationKeys)
+    {
+        $translationKeys[] = 'JsTrackerInstallCheck_TestInstallationDescription';
+        $translationKeys[] = 'JsTrackerInstallCheck_TestInstallationBtnText';
+        $translationKeys[] = 'JsTrackerInstallCheck_JsTrackingCodeInstallCheckSuccessMessage';
+        $translationKeys[] = 'JsTrackerInstallCheck_JsTrackingCodeInstallCheckFailureMessage';
+        $translationKeys[] = 'General_Testing';
+    }
+
+    public function getStylesheetFiles(&$stylesheets)
+    {
+        $stylesheets[] = "plugins/JsTrackerInstallCheck/vue/src/JsTrackerInstallCheck/JsTrackerInstallCheck.less";
+    }
+
+    public function isExcludedVisit(&$excluded, Request $request)
+    {
+        // We don't have an early return when the request has already been excluded because we want the test to work even if the request is excluded due to a VPN or something
+
+        $hasInstallCheckParam = $request->hasParam(self::QUERY_PARAM_NAME);
+        if (!$hasInstallCheckParam) {
+            return;
+        }
+
+        $trackerInstallCheckParam = $request->getParams()[self::QUERY_PARAM_NAME];
+        if (empty($trackerInstallCheckParam)) {
+            return;
+        }
+
+        // If the request has already been excluded, we don't want to override that
+        if (empty($excluded)) {
+            $excluded = 'excluded: testing whether JS tracker is installed';
+        }
```

Comment:
`$excluded` was actually meant to be a boolean. At least that is what the event description says. Providing any string reason won't make a difference, as the value is used for comparing in core only. (The TrackingSpamPrevention plugin might use this incorrectly, though.)
Thus you actually don't need to check the value before. As it should always be `true` in the end.
```suggestion
        $excluded = true;
```

Reply:
Yeah. I was copying TrackingSpamPrevention when I should have checked the event implementation. I'll clean this up.